### ISSUE TITLE
Random module functions slowdown on multiple cores

### ISSUE THREAD

**Issue by Sudha247** [ocaml-multicore/ocaml-multicore#333]
The code snippet creates a float array and assigns a random value to all the elements: 

```ocaml
module T = Domainslib.Task

let num_domains = try int_of_string Sys.argv.(1) with _ -> 4
let n = try int_of_string Sys.argv.(2) with _ -> 10000
let cs = try int_of_string Sys.argv.(3) with _ -> 32

let _ =
  let arr = Array.create_float n in
  (*init with random number*)
  let pool = T.setup_pool ~num_domains:(num_domains - 1) in
  T.parallel_for pool ~chunk_size:(n/num_domains) ~start:0 ~finish:(pred n) 
    ~body:(fun i -> Array.unsafe_set arr i (Random.float 100.));
  T.teardown_pool

```

There is a significant slowdown as the number of cores increase. For input `100_000_000` the following numbers were observed on `parallel_minor_gc`:

| Cores | time(s) |
|-------|---------|
| 1     | 2.414   |
| 2     | 5.383   |
| 4     | 5.795   |

Perf report indicates overhead at `Random` as 92.07% on 4 cores. 

![image](https://user-images.githubusercontent.com/13328130/80487890-27352b00-897b-11ea-8738-ad70b9e6a9a9.png)

**Comment by stedolan** [comment]
I think this might be because all calls to Random share common state. Have you tried using per-domain or per-task instance of `Random.State.make_self_init ()`? That way, you can have many different states with no sharing.

**Comment by Sudha247** [comment]
Thanks for the pointer @stedolan. I tried a per-domain instance of `Random.State.make_self_init ()`, here's the code (I'm not sure if this is the best approach):

```ocaml
module T = Domainslib.Task
let n = try int_of_string Sys.argv.(2) with _ -> 1000
let num_domains = try int_of_string Sys.argv.(1) with _ -> 4

let arr = Array.create_float n

let _ =
  let domains = T.setup_pool ~num_domains:(num_domains - 1) in
  let states = Array.init num_domains (fun _ -> Random.State.make_self_init()) in
  T.parallel_for domains ~chunk_size:(n/num_domains) ~start:0 ~finish:(n-1)
  ~body:(fun i ->
    let d = (Domain.self() :> int) mod num_domains in
    Array.unsafe_set arr i (Random.State.float states.(d) 100. ))
```
I did not observe much slowdown in this version, and there was some amount of speedup as well. There's still 60-80% overhead at `Random` and as per my observation, the overhead increases as the input size increases. These are some numbers for input `10_000_000`. Would appreciate insights on the code and how to improve it.

| #cores | time (s) |
|--------|----------|
| 1      | 0.727    |
| 2      | 0.537    |
| 4      | 0.630    |
| 8      | 0.393    |
| 12     | 0.254    |
| 16     | 0.243    |
| 20     | 0.220    |
| 24     | 0.161    |

**Comment by stedolan** [comment]
Since all of the Randoms are allocated at the same time by the main domain, it's possible that they share cache lines. This will cause contention when they're used by multiple domains, because multiple domains will use the same cache line even though they're using different parts of it (something known as "false sharing").

Have you tried creating the Randoms on the domain that's going to use them? That should avoid the false sharing as different domains will allocate memory that's far apart. (I'm not sure how this fits into the setup_pool API though)

**Comment by bluddy** [comment]
I modified the code above to also include a multiplier, spacing out the array so that cache lines don't cross:

```ocaml
module T = Domainslib.Task
let n = try int_of_string Sys.argv.(2) with _ -> 1000
let num_domains = try int_of_string Sys.argv.(1) with _ -> 4
let mult = try int_of_string Sys.argv.(3) with _ -> 1

let arr = Array.create_float n

let _ =
  let domains = T.setup_pool ~num_domains:(num_domains - 1) in
  let states = Array.init (num_domains * mult) (fun _ -> Random.State.make_self_init()) in
  T.parallel_for domains ~chunk_size:(n/num_domains) ~start:0 ~finish:(n-1)
  ~body:(fun i ->
    let d = (Domain.self() :> int) mod num_domains in
    Array.unsafe_set arr i (Random.State.float states.(d * mult) 100. ))
```

There's a slightly higher cost for initializing more states, but it's normally negligible. Results for `10_000_000`:

Domains | Multiplier | Time (s)
--- | --- | ---
1 | 1 | 0.476
1 | 2 | 0.464
1 | 4 | 0.479
2 | 1 | 0.370
2 | 2 | 0.254
2 | 4 | 0.245
4 | 1 | 0.351
4 | 2 | 0.245
4 | 4 | 0.163
4 | 8 | 0.170
8 | 1 | 0.255
8 | 2 | 0.164
8 | 4 | 0.144

Note that this is on my laptop that has 4 hyperthreaded cores, so you don't get any huge benefit beyond 4 cores.

**Comment by Sudha247** [comment]
Here's one way to create Random states on the domain that is going to use them as Stephen suggested, to avoid false sharing.

```ocaml
module T = Domainslib.Task
let n = try int_of_string Sys.argv.(2) with _ -> 1000
let num_domains = try int_of_string Sys.argv.(1) with _ -> 4

let arr = Array.create_float n

let init_part s e arr =
    let my_state = Random.State.make_self_init () in
    for i = s to e do
      Array.unsafe_set arr i (Random.State.float my_state 100.)
    done

let _ =
  let domains = T.setup_pool ~num_domains:(num_domains - 1) in
  T.parallel_for domains ~chunk_size:1 ~start:0 ~finish:(num_domains - 1)
  ~body:(fun i -> init_part (i * n / num_domains) ((i+1) * n / num_domains - 1) arr);
  T.teardown_pool domains
```
This seems to scale reasonably. Numbers for input `100_000_000`:
| #Cores | Time (s) |
|--------|----------|
| 1      | 3.035    |
| 2      | 1.539    |
| 4      | 0.790    |
| 8      | 0.417    |
| 12     | 0.295    |
| 16     | 0.238    |
| 20     | 0.203    |
| 24     | 0.180    |

**Comment by bluddy** [comment]
I'm not sure moving the chunking to the user code is cleaner than creating an artificially bigger shared state. A `parallel_for_init` function that had an `init` lambda argument and fed the result to the `body` could work well in this case.

**Comment by stedolan** [comment]
This will be a common enough problem that there should definitely be something in domainslib to handle it.

@bluddy Thanks for the updated code and benchmarks, but I think your new code might be a bit misleading. I don't think making the `states` array bigger helps at all here.

In the original code, the different domains are accessing different fields of a short array, so they all access the same cacheline. There is no performance problem here, because they only read the array. The cacheline quickly ends up shared between all cores.

The contention is not on the array, but on the Random.State objects pointed to by the array. @Sudha247's code allocates those all at once on a single domain. The allocator returns allocations that are close together, so these end up sharing cachelines. It is these lines that are contended.

Your code uses a larger array, initialised with `Array.init (num_domains * mult) (fun _ -> Random.State.make_self_init())`. This means that between the allocations of random states that are actually used, there are `mult - 1` unused states allocated. These unused states occupy enough space that the useful allocations are on separate cache lines.

I think you'd see the same results if you kept the original short array, removed all the multiplication, but initialised the array like this:

    let states = Array.init num_domains (fun _ ->
         for i = 1 to mult do ignore (Random.State.make_self_init ()) done;
         Random.State.make_self_init ())

I think it's still better to allocate the state on the domain that will actually use it, though.

**Comment by bluddy** [comment]
> The contention is not on the array, but on the Random.State objects pointed to by the array. @Sudha247's code allocates those all at once on a single domain. The allocator returns allocations that are close together, so these end up sharing cachelines. It is these lines that are contended.

Yep, agreed. That's exactly what I meant above.

> I think it's still better to allocate the state on the domain that will actually use it, though.

True, but I don't particularly like @Sudha247's approach in the latter example of making the user responsible for the chunking. The best I can think of is having a `parallel_for_init` function:

```ocaml
val parallel_for_init: int -> init:(unit -> 'a) -> chunk_size:int -> start:int -> finish:int -> (int -> 'a -> ())
```

Which will run init before running the parallel for and then pass the data to the lambda. However, this still doesn't address initialization of long-term storage that's too expensive to run on every `parallel_for` execution. That's still done most naturally at the master domain.

**Comment by bluddy** [comment]
Another, possibly simpler solution is to just have a parallel_for that goes and does the initialization in the separate domains. Since the array points to the data structures, it should be fine. Haven't tested the code below, but it should work:

```ocaml
module T = Domainslib.Task
let n = try int_of_string Sys.argv.(2) with _ -> 1000
let num_domains = try int_of_string Sys.argv.(1) with _ -> 4

let arr = Array.create_float n

let _ =
  let domains = T.setup_pool ~num_domains:(num_domains - 1) in
  (* We have to initialize with something *)
  let states = Array.init num_domains (fun _ -> Random.State.make_self_init()) in
  (* Overwrite using separate domains *)
  T.parallel_for domains ~chunk_size:1 ~start:0 ~finish:(num_domains-1)
  ~body:(fun i -> Array.unsafe_set states i (Random.State.make_self_init()));
  T.parallel_for domains ~chunk_size:(n/num_domains) ~start:0 ~finish:(n-1)
  ~body:(fun i ->
    let d = (Domain.self() :> int) mod num_domains in
    Array.unsafe_set arr i (Random.State.float states.(d) 100. ))
```

**Comment by Sudha247** [comment]
@bluddy thanks for the idea and code, but I don't think this one scales better than other versions.

```ocaml
T.parallel_for domains ~chunk_size:1 ~start:0 ~finish:(num_domains-1)
  ~body:(fun i -> Array.unsafe_set states i (Random.State.make_self_init()));
```
While one might expect this `parallel_for` to run exactly one chunk in every domain, that need not necessarily happen every time. This is due to the fact that the Task API is designed in such a way where any domain that is free would take up the next available task. Since the task in this case is quite small and takes very little time to execute, the domains might be freed soon enough to take up next available task, meaning one domain might not execute any task at all and some other domain might execute it twice. I am not sure if this was an assumption you made, just putting this out here.

There's quite a bit of false sharing happening in your latest code too. Here are the stats on 4 cores:
```
Shared Data Cache Line Table     (3 entries, sorted on Total HITMs)
       ----------- Cacheline ----------    Total      Tot  ----- LLC Load Hitm -----  ---- Store Reference ----  --- Load Dram ----
Index             Address  Node  PA cnt  records     Hitm    Total      Lcl      Rmt    Total    L1Hit   L1Miss       Lcl       Rmt
    0      0x7f0b9330fdc0     0   16701    18705   94.44%     1715     1715        0     2337      795     1542         0         0
    1      0x7f0b932dfb80     0     348      475    5.40%       98       98        0      155      103       52         0         0
    2  0xffff9854071bd5c0     0       1        2    0.11%        2        2        0        0        0        0         0         0
```
I had tried something similar earlier; I created an empty hash table and added one Random state each from every spawned domain with the Domain id as key. The results were not very convincing and false sharing was happening in that version as well.

**Comment by bluddy** [comment]
Hmm... That's disappointing. I was indeed assuming every domain would get one task to execute. Would it be possible to add an argument to the API that forces uniqueness of domain? I could imagine that being valuable in other contexts.

On the other hand, trying to get different domain allocators to allocate feels... unnatural on some level. There's no awareness at the language level of these separate allocators, and they seem very implementation-specific (for example, I'm not sure the same solution would work in webassembly). In C, I would just use gaps in data structures for this niche purpose of falsely shared small mutable state, and I also believe that's the standard solution.

**Comment by kayceesrk** [comment]
> On the other hand, trying to get different domain allocators to allocate feels... unnatural on some level. 

I’d disagree on this point. On Linux, threads allocate in their NUMA domains by default. This helps avoid cross socket memory access whose bandwidth is limited compared to accessing on the same socket. Hence, allocating by the same domain is the most natural thing to do.

Also while one could argue that this is low level detail not exposed by the language, what we are discussing here is performance and not correctness. For performance in OCaml, the knowledge of the underlying organisation of the heap is certainly important.

**Comment by kayceesrk** [comment]
Fixed by #372 Domain-local storage. See impressive speedups in https://github.com/ocaml-multicore/ocaml-multicore/issues/336 and https://github.com/ocaml-bench/sandmark/pull/157.
